managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java

```diff
@@ -268,15 +277,53 @@ protected void recoverFromLedger(final ManagedCursorInfo info, final VoidCallbac
                 }
 
                 PositionImpl position = new PositionImpl(positionInfo);
+                PositionImpl lastIndividualAckedMessage = null;
                 if (positionInfo.getIndividualDeletedMessagesCount() > 0) {
                     recoverIndividualDeletedMessages(positionInfo.getIndividualDeletedMessagesList());
+                    // also recover readPosition to last acked message
+                    lastIndividualAckedMessage = new PositionImpl(
+                            positionInfo.getIndividualDeletedMessagesList().get(positionInfo.getIndividualDeletedMessagesCount() - 1).getUpperEndpoint());
                 }
                 recoveredCursor(position);
+                // if individualDeletedMessages are recovered then readPosition should be set after last Acked-Message
+                if (lastIndividualAckedMessage != null) {
+                    this.readPosition = lastIndividualAckedMessage;
+                }
                 callback.operationComplete();
             }, null);
         }, null);
     }
 
+    /**
+     * return unacked messages which are not part of individual deleted messages
+     * 
+     * @return
+     */
+    @Override
+    public Set<? extends Position> getNotDeletedMessages() {
```

Comment:
I don't think we need to go down this route. 

When reading, the managed cursor is already skipping all the individually deleted messages. We should just set the read position to be always on (mark-delete + 1) when recovering.